src/server/qtquick/private/wquickinputmethodv2.cpp

```diff
@@ -0,0 +1,473 @@
+    return d->handle->delete_c.after_length;
+}
+
+QString WInputMethodV2State::preeditStringText() const
+{
+    W_DC(WInputMethodV2State);
+    return d->handle->preedit.text;
+}
+
+qint32 WInputMethodV2State::preeditStringCursorBegin() const
+{
+    W_DC(WInputMethodV2State);
+    return d->handle->preedit.cursor_begin;
+}
+
+qint32 WInputMethodV2State::preeditStringCursorEnd() const
+{
+    W_DC(WInputMethodV2State);
+    return d->handle->preedit.cursor_end;
+}
+
+WInputMethodV2State::WInputMethodV2State(wlr_input_method_v2_state *handle, QObject *parent)
+    : QObject(parent)
+    , WObject(*new WInputMethodV2StatePrivate(handle, this))
+{}
+
+class WInputPopupSurfacePrivate : public WObjectPrivate
+{
+public:
+    W_DECLARE_PUBLIC(WInputPopupSurface)
+    explicit WInputPopupSurfacePrivate(WQuickInputPopupSurfaceV2 *surface, WInputPopupSurface *qq)
+        : WObjectPrivate(qq)
+        , handle(surface)
+    { }
+    WQuickInputPopupSurfaceV2 *handle;
+};
+
+WInputPopupSurface::WInputPopupSurface(WQuickInputPopupSurfaceV2 *surface, QObject *parent)
+    : WToplevelSurface(parent)
+    , WObject(*new WInputPopupSurfacePrivate(surface, this))
+{
+    W_D(WInputPopupSurface);
+    connect(d->handle, &WQuickInputPopupSurfaceV2::beforeDestroy, this, [this]() { Q_EMIT this->beforeDestroy(this); this->deleteLater(); });
+}
+
+bool WInputPopupSurface::doesNotAcceptFocus() const
+{
+    return true;
+}
+
+WSurface *WInputPopupSurface::surface() const
+{
+    W_DC(WInputPopupSurface);
+    return d->handle->surface();
+}
+
+bool WInputPopupSurface::isActivated() const
+{
+    return true;
+}
+
+QRect WInputPopupSurface::getContentGeometry() const
+{
+    auto s = this->surface()->handle()->handle();
+    return {s->current.dx, s->current.dy, s->current.width, s->current.height};
+}
+
+bool WInputPopupSurface::checkNewSize(const QSize &size)
+{
+    auto s = this->surface()->handle()->handle();
+    QSize current(s->current.width, s->current.height);
+    return (size == current);
+}
+
+WInputPopupSurfaceItem::WInputPopupSurfaceItem(QQuickItem *parent)
+    : WSurfaceItem(parent)
+    , m_inputPopupSurface(nullptr)
+{}
+
+WInputPopupSurface *WInputPopupSurfaceItem::surface() const
+{
+    return m_inputPopupSurface;
+}
+
+void WInputPopupSurfaceItem::setSurface(WInputPopupSurface *surface)
+{
+    if (m_inputPopupSurface == surface)
+        return;
+    m_inputPopupSurface = surface;
+    WSurfaceItem::setSurface(surface ? surface->surface() : nullptr);
+    Q_EMIT this->surfaceChanged();
+}
+
+QPointF WInputPopupSurfaceItem::implicitPosition() const
+{
+    return QPointF(0, 0);
+}
+
+QSize WInputPopupSurfaceItem::minimumSize() const
+{
+    return QSize();
+}
+
+QSize WInputPopupSurfaceItem::maximumSize() const
+{
+    return QSize();
+}
+
+void WInputPopupSurfaceItem::onSurfaceCommit()
+{
+    auto wlrSurface = m_inputPopupSurface->surface()->handle()->handle();
+    qCDebug(qLcInputMethod) << "Surface item committed."
```

Comment:
This function will be call by each repaint from input method window, it is not suggest to print log here.

Reply:
This is just for debugging and will be removed when the draft is ready.